### PR TITLE
Delete Device2Host caused by comm with device and host

### PR DESCRIPTION
**Pull request by zhaozheng09**
## Delete some D2H ops

`if not` will be bring a sync within host and device, so replace it from `if not` to `torch.where`

<details>
  <summary>Before submitting</summary>

- [ ] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [ ] Did you write any new **necessary tests**?

</details>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

before:
image
after:
image
Host will be block until device done, this D2H maybe block pipeline of host and device, so I use `where op` to replace `if not`

### CODE REVIEW

**Review by Borda — COMMENTED** [review]
can we pls add tests for this fix?

**Comment by Borda** [comment]
> this code will raise in old torch version, I don't know how to cover, please give me an example ? thx very much

ok, so you say that this PR does not have any impact on functional metrics results but improves performances, correct?

**Review by Borda — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]

**Comment by SkafteNicki** [comment]
did a couple of changes to the PR:
* move the logic `torch.where` logic to a separate function such that it could be used multiple places
* made the testing of cpu+sigmoid+half less restrictive e.g. instead of always being skipped it is now only skipped for older pytorch builds

File: src/torchmetrics/utilities/compute.py [hunk 1]
```diff
@@ -184,3 +185,45 @@ def interp(x: Tensor, xp: Tensor, fp: Tensor) -> Tensor:
...
+        tensor([[0.0900, 0.2447, 0.6652],
+                [0.6652, 0.2447, 0.0900]])
+        >>> tensor = torch.tensor([0.0, 0.5, 1.0])
+        >>> normalize_logits_if_needed(tensor, normalization="sigmoid")
+        tensor([0.0000, 0.5000, 1.0000])
+
+    """
+    # decrease sigmoid on cpu .
+    if tensor.device == torch.device("cpu"):
+        if not torch.all((preds >= 0) * (preds <= 1)):
```
Comment (Borda) [inline src/torchmetrics/utilities/compute.py:219]: `preds` is not defined...

**Review by justusschock — APPROVED** [review]

File: CHANGELOG.md [hunk 2]
```diff
@@ -27,6 +27,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Fixed
 
+- Delete `Device2Host` caused by comm with device and host ([#2840](https://github.com/PyTorchLightning/metrics/pull/2840))
```
Comment (Borda) [inline CHANGELOG.md:30]:
```suggestion
- Delete `Device2Host` caused by comm with device and host ([#2840](https://github.com/PyTorchLightning/metrics/pull/2840))


```

**Review by lantiga — APPROVED** [review]
Looks good, thanks!

In general we will have to become sensitive to data-dependent control flow, to make torchmetrics more compiler-friendly in the future. This applies to both before and after this PR of course, having the condition in a separate function like it is now allows us to side-step the problem, so the final state is better from this point of view.
